File: src/markdown.ts
// Markdown → Google Docs API requests, and Google Doc HTML → Markdown

interface InlineSegment {
  text: string;
  bold?: boolean;
  italic?: boolean;
}

interface DocSegment {
  inlines: InlineSegment[];
  style: "HEADING_1" | "HEADING_2" | "HEADING_3" | "NORMAL_TEXT";
  bullet?: boolean;
}

function parseInline(text: string): InlineSegment[] {
  const result: InlineSegment[] = [];
  let i = 0;
  let current = "";

  while (i < text.length) {
    if (text[i] === "*" && text[i + 1] === "*") {
      if (current) result.push({ text: current });
      current = "";
      i += 2;
      let bold = "";
      while (i < text.length && !(text[i] === "*" && text[i + 1] === "*")) {
        bold += text[i++];
      }
      i += 2;
      result.push({ text: bold, bold: true });
    } else if (text[i] === "*" && text[i + 1] !== "*") {
      if (current) result.push({ text: current });
      current = "";
      i++;
      let italic = "";
      while (i < text.length && text[i] !== "*") {
        italic += text[i++];
      }
      i++;
      result.push({ text: italic, italic: true });
    } else if (text[i] === "`") {
      if (current) result.push({ text: current });
      current = "";
      i++;
      let code = "";
      while (i < text.length && text[i] !== "`") {
        code += text[i++];
      }
      i++;
      result.push({ text: code, italic: true });
    } else {
      current += text[i++];
    }
  }

  if (current) result.push({ text: current });
  return result;
}

export function markdownToDocRequests(markdown: string): object[] {
  const lines = markdown.split("\n");
  const segments: DocSegment[] = [];

  for (const line of lines) {
    const h1 = line.match(/^# (.+)/);
    const h2 = line.match(/^## (.+)/);
    const h3 = line.match(/^### (.+)/);
    const bullet = line.match(/^[-*] (.+)/);

    if (h1) {
      segments.push({ inlines: [{ text: h1[1] }], style: "HEADING_1" });
    } else if (h2) {
      segments.push({ inlines: [{ text: h2[1] }], style: "HEADING_2" });
    } else if (h3) {
      segments.push({ inlines: [{ text: h3[1] }], style: "HEADING_3" });
    } else if (bullet) {
      segments.push({
        inlines: parseInline(bullet[1]),
        style: "NORMAL_TEXT",
        bullet: true,
      });
    } else if (line.trim() === "") {
      segments.push({ inlines: [{ text: "" }], style: "NORMAL_TEXT" });
    } else {
      segments.push({ inlines: parseInline(line), style: "NORMAL_TEXT" });
    }
  }

  // Build full text and track index ranges
  const requests: object[] = [];
  const styleRanges: Array<{
    startIndex: number;
    endIndex: number;
    paragraphStyle?: string;
    textStyle?: { bold?: boolean; italic?: boolean };
    bullet?: boolean;
  }> = [];

  let fullText = "";
  let currentIndex = 1;

  for (const segment of segments) {
    const segStart = currentIndex;
    const segText =
      segment.inlines.map((i) => i.text).join("") + "\n";

    fullText += segText;

    // Track paragraph style range
    if (segment.style !== "NORMAL_TEXT" || segment.bullet) {
      styleRanges.push({
        startIndex: segStart,
        endIndex: segStart + segText.length,
        paragraphStyle: segment.bullet ? undefined : segment.style,
        bullet: segment.bullet,
      });
    }

    // Track inline style ranges
    let inlineIndex = segStart;
    for (const inline of segment.inlines) {
      if (inline.bold || inline.italic) {
        styleRanges.push({
          startIndex: inlineIndex,
          endIndex: inlineIndex + inline.text.length,
          textStyle: { bold: inline.bold, italic: inline.italic },
        });
      }
      inlineIndex += inline.text.length;
    }

    currentIndex += segText.length;
  }

  // Single text insertion
  if (fullText) {
    requests.push({
      insertText: {
        location: { index: 1 },
        text: fullText,
      },
    });
  }

  // Apply styles
  for (const range of styleRanges) {
    if (range.paragraphStyle) {
      requests.push({
        updateParagraphStyle: {
          range: { startIndex: range.startIndex, endIndex: range.endIndex },
          paragraphStyle: { namedStyleType: range.paragraphStyle },
          fields: "namedStyleType",
        },
      });
    }
    if (range.bullet) {
      requests.push({
        createParagraphBullets: {
          range: { startIndex: range.startIndex, endIndex: range.endIndex },
          bulletPreset: "BULLET_DISC_CIRCLE_SQUARE",
        },
      });
    }
    if (range.textStyle) {
      requests.push({
        updateTextStyle: {
          range: { startIndex: range.startIndex, endIndex: range.endIndex },
          textStyle: range.textStyle,
          fields: Object.keys(range.textStyle).join(","),
        },
      });
    }
  }

  return requests;
}

File: src/kuali_docs_mcp/markdown.py
"""Markdown → Google Docs API batchUpdate requests."""

from dataclasses import dataclass, field


@dataclass
class InlineSegment:
    text: str
    bold: bool = False
    italic: bool = False


@dataclass
class DocSegment:
    inlines: list[InlineSegment]
    style: str = "NORMAL_TEXT"  # HEADING_1, HEADING_2, HEADING_3, NORMAL_TEXT
    bullet: bool = False


def _parse_inline(text: str) -> list[InlineSegment]:
    result = []
    i = 0
    current = ""

    while i < len(text):
        if text[i] == "*" and i + 1 < len(text) and text[i + 1] == "*":
            if current:
                result.append(InlineSegment(current))
            current = ""
            i += 2
            bold = ""
            while i < len(text) and not (text[i] == "*" and i + 1 < len(text) and text[i + 1] == "*"):
                bold += text[i]
                i += 1
            i += 2
            result.append(InlineSegment(bold, bold=True))
        elif text[i] == "*" and (i + 1 >= len(text) or text[i + 1] != "*"):
            if current:
                result.append(InlineSegment(current))
            current = ""
            i += 1
            italic = ""
            while i < len(text) and text[i] != "*":
                italic += text[i]
                i += 1
            i += 1
            result.append(InlineSegment(italic, italic=True))
        elif text[i] == "`":
            if current:
                result.append(InlineSegment(current))
            current = ""
            i += 1
            code = ""
            while i < len(text) and text[i] != "`":
                code += text[i]
                i += 1
            i += 1
            result.append(InlineSegment(code, italic=True))
        else:
            current += text[i]
            i += 1

    if current:
        result.append(InlineSegment(current))
    return result


def _parse_markdown(markdown: str) -> list[DocSegment]:
    segments = []
    for line in markdown.splitlines():
        if line.startswith("# "):
            segments.append(DocSegment([InlineSegment(line[2:])], style="HEADING_1"))
        elif line.startswith("## "):
            segments.append(DocSegment([InlineSegment(line[3:])], style="HEADING_2"))
        elif line.startswith("### "):
            segments.append(DocSegment([InlineSegment(line[4:])], style="HEADING_3"))
        elif line.startswith("- ") or line.startswith("* "):
            segments.append(DocSegment(_parse_inline(line[2:]), bullet=True))
        elif line.strip() == "":
            segments.append(DocSegment([InlineSegment("")]))
        else:
            segments.append(DocSegment(_parse_inline(line)))
    return segments


def markdown_to_doc_requests(markdown: str) -> list[dict]:
    segments = _parse_markdown(markdown)
    requests = []
    style_ranges = []

    full_text = ""
    current_index = 1

    for segment in segments:
        seg_start = current_index
        seg_text = "".join(s.text for s in segment.inlines) + "\n"
        full_text += seg_text

        if segment.style != "NORMAL_TEXT":
            style_ranges.append({
                "type": "paragraph",
                "startIndex": seg_start,
                "endIndex": seg_start + len(seg_text),
                "style": segment.style,
            })

        if segment.bullet:
            style_ranges.append({
                "type": "bullet",
                "startIndex": seg_start,
                "endIndex": seg_start + len(seg_text),
            })

        inline_index = seg_start
        for inline in segment.inlines:
            if inline.bold or inline.italic:
                style_ranges.append({
                    "type": "text",
                    "startIndex": inline_index,
                    "endIndex": inline_index + len(inline.text),
                    "bold": inline.bold,
                    "italic": inline.italic,
                })
            inline_index += len(inline.text)

        current_index += len(seg_text)

    if full_text:
        requests.append({"insertText": {"location": {"index": 1}, "text": full_text}})

    for r in style_ranges:
        if r["type"] == "paragraph":
            requests.append({
                "updateParagraphStyle": {
                    "range": {"startIndex": r["startIndex"], "endIndex": r["endIndex"]},
                    "paragraphStyle": {"namedStyleType": r["style"]},
                    "fields": "namedStyleType",
                }
            })
        elif r["type"] == "bullet":
            requests.append({
                "createParagraphBullets": {
                    "range": {"startIndex": r["startIndex"], "endIndex": r["endIndex"]},
                    "bulletPreset": "BULLET_DISC_CIRCLE_SQUARE",
                }
            })
        elif r["type"] == "text":
            fields = ",".join(k for k in ("bold", "italic") if r.get(k))
            requests.append({
                "updateTextStyle": {
                    "range": {"startIndex": r["startIndex"], "endIndex": r["endIndex"]},
                    "textStyle": {k: r[k] for k in ("bold", "italic") if r.get(k)},
                    "fields": fields,
                }
            })

    return requests
